File: packages/guards/src/phone-format.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type PhoneRegion = 'us' | 'kr' | 'jp' | 'cn' | 'uk' | 'international';

interface PhoneFormatOptions {
  action: 'block' | 'warn' | 'mask';
  regions?: PhoneRegion[];
}

const PHONE_PATTERNS: Record<PhoneRegion, RegExp> = {
  us: /\b(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b/g,
  kr: /\b(?:\+?82[-.\s]?)?0?\d{2,3}[-.\s]?\d{3,4}[-.\s]?\d{4}\b/g,
  jp: /\b(?:\+?81[-.\s]?)?0?\d{1,4}[-.\s]?\d{1,4}[-.\s]?\d{4}\b/g,
  cn: /\b(?:\+?86[-.\s]?)?1[3-9]\d{9}\b/g,
  uk: /\b(?:\+?44[-.\s]?)?0?\d{4}[-.\s]?\d{6}\b/g,
  international: /\+\d{1,3}[-.\s]?\d{1,14}/g,
};

interface PhoneMatch {
  phone: string;
  region: PhoneRegion;
  start: number;
  end: number;
}

const ALL_REGIONS: PhoneRegion[] = ['us', 'kr', 'jp', 'cn', 'uk', 'international'];

function detect(text: string, regions: PhoneRegion[]): PhoneMatch[] {
  const matches: PhoneMatch[] = [];
  const seen = new Set<string>();

  for (const region of regions) {
    const re = new RegExp(PHONE_PATTERNS[region].source, 'g');
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      const key = `${m.index}-${m[0].length}`;
      if (seen.has(key)) continue;
      seen.add(key);
      matches.push({
        phone: m[0],
        region,
        start: m.index,
        end: m.index + m[0].length,
      });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function maskPhones(text: string, matches: PhoneMatch[]): string {
  let result = text;
  for (const m of matches) {
    result =
      result.slice(0, m.start) + '[PHONE]' + result.slice(m.end);
  }
  return result;
}

export function phoneFormat(options: PhoneFormatOptions): Guard {
  const regions = options.regions ?? ALL_REGIONS;

  return {
    name: 'phone-format',
    version: '0.1.0',
    description: 'International phone number detection and masking',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, regions);
      const triggered = matches.length > 0;

      if (!triggered) {
        return {
          guardName: 'phone-format',
          passed: true,
          action: 'allow',
          latencyMs: Math.round(performance.now() - start),
        };
      }

      if (options.action === 'mask') {
        return {
          guardName: 'phone-format',
          passed: true,
          action: 'override',
          overrideText: maskPhones(text, matches),
          latencyMs: Math.round(performance.now() - start),
          details: {
            detected: matches.map(({ phone, region }) => ({
              phone, region,
            })),
          },
        };
      }

      return {
        guardName: 'phone-format',
        passed: false,
        action: options.action,
        latencyMs: Math.round(performance.now() - start),
        details: {
          detected: matches.map(({ phone, region }) => ({
            phone, region,
          })),
        },
      };
    },
  };
}

File: python/open_guardrail/guards/phone_format.py
"""Phone number format detection guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "international": re.compile(
        r"\+\d{1,3}[\s-]?\d[\d\s-]{4,14}\d"
    ),
    "us": re.compile(r"\d{3}-\d{3}-\d{4}"),
}

_MASK_LABELS: dict[str, str] = {
    "international": "[PHONE]",
    "us": "[PHONE]",
}


def _detect(
    text: str, entities: list[str]
) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append(
                {
                    "type": entity,
                    "value": m.group(),
                    "start": m.start(),
                    "end": m.end(),
                }
            )
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask_text(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        label = _MASK_LABELS.get(m["type"], "[PII]")
        result = (
            result[: m["start"]] + label + result[m["end"] :]
        )
    return result


class _PhoneFormat:
    def __init__(
        self,
        *,
        action: str = "block",
        entities: Optional[List[str]] = None,
    ) -> None:
        self.name = "phone-format"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000

        if not matches:
            return GuardResult(
                guard_name="phone-format",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        detected = [
            {"type": m["type"], "value": m["value"]}
            for m in matches
        ]

        if self.action == "mask":
            return GuardResult(
                guard_name="phone-format",
                passed=True,
                action="override",
                override_text=_mask_text(text, matches),
                latency_ms=round(elapsed, 2),
                details={"detected": detected},
            )

        return GuardResult(
            guard_name="phone-format",
            passed=False,
            action=self.action,
            message="Phone number detected in text",
            latency_ms=round(elapsed, 2),
            details={
                "detected": detected,
                "reason": (
                    "Text contains phone numbers"
                ),
            },
        )


def phone_format(
    *,
    action: str = "block",
    entities: Optional[List[str]] = None,
) -> _PhoneFormat:
    return _PhoneFormat(action=action, entities=entities)
